nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java

```diff
@@ -0,0 +1,286 @@
+                    .name("Max Row Number")
+                    .description(
+                            "Specifies the max row number of each tablet. It can be updated by expression language.")
+                    .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+                    .defaultValue("false")
+                    .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+                    .required(false)
+                    .build();
+
+    static {
+        descriptors.add(RECORD_READER_FACTORY);
+        descriptors.add(TIME_FIELD);
+        descriptors.add(SCHEMA);
+        descriptors.add(ALIGNED);
+        descriptors.add(MAX_ROW_NUMBER);
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext processContext, ProcessSession processSession)
+            throws ProcessException {
+        final RecordReaderFactory recordParserFactory =
+                processContext
+                        .getProperty(RECORD_READER_FACTORY)
+                        .asControllerService(RecordReaderFactory.class);
+
+        String timeFieldProperty = processContext.getProperty(TIME_FIELD).getValue();
+        String schemaProperty = processContext.getProperty(SCHEMA).getValue();
+        String alignedProperty = processContext.getProperty(ALIGNED).getValue();
+        String maxRowNumberProperty = processContext.getProperty(MAX_ROW_NUMBER).getValue();
+
+        FlowFile flowFile = processSession.get();
+
+        String timeField = timeFieldProperty != null ? timeFieldProperty : "Time";
+        Boolean aligned = alignedProperty != null ? Boolean.valueOf(alignedProperty) : false;
+        int maxRowNumber = maxRowNumberProperty != null ? Integer.valueOf(maxRowNumberProperty) : 1024;
+
+        try (final InputStream inputStream = processSession.read(flowFile);
+             final RecordReader recordReader =
+                     recordParserFactory.createRecordReader(flowFile, inputStream, getLogger())) {
+            if (flowFile == null) {
+                inputStream.close();
+                recordReader.close();
+                processSession.transfer(flowFile, REL_SUCCESS);
+                return;
+            }
+
+            HashMap<String, Tablet> tablets;
+            boolean needInitFormatter;
+            Schema schema;
+            Tuple<Boolean, String> result;
+
+            result =
+                    schemaProperty != null
+                            ? validateSchemaAttribute(schemaProperty)
+                            : validateSchema(recordReader.getSchema());
+
+            if (!result.getKey()) {
+                getLogger().error(result.getValue());
+                inputStream.close();
+                recordReader.close();
```

Comment:
Closing these objects should be unnecessary with the use of try-with-resources.

Reply:
I have tried to delete these, but it will throw an exception `java.lang.IllegalStateException: FlowFile[0,16972544312800.mockFlowFile,0B] already in use for an active callback or InputStream created by ProcessSession.read(FlowFile) has not been closed`.

Reply:
Thanks for the reply @xuanronaldo, does this occur when calling `ProcessSession.transfer()`? If you move it outside of the try-catch block, or change the `read()` signature to use a callback, that should resolve this error.